### PR TITLE
WIP Made initial changes to enable dimension properties and added attention module

### PR DESCRIPTION
**Pull request by vincentpierre**
### Proposed change(s)

Implementation according to this design : https://docs.google.com/document/d/1EvgrFjICvz9o3ymAVnR1O7nNxeH-4UknahQTGcjG3ww/edit#

When trying to train with a BufferSensor, the user will see the error message :
```
mlagents.trainers.exception.UnityTrainerException: Unsupported shape of (A, B) for observation i
```
This Pr is just setting the ground work for allowing variable size observation (no trainer changes)

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: com.unity.ml-agents/Runtime/Sensors/IDimensionPropertiesSensor.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        /// <summary>
+        /// Means it is possible to do a convolution in this dimension.
+        /// </summary>
+        TranslationalEquivariance = 2,
+
+        /// <summary>
+        /// Means that there can be a variable number of observations in this dimension.
+        /// The observations are unordered.
+        /// </summary>
+        VariableSize = 3,
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/IDimensionPropertiesSensor.cs:30]: These are flags so the values should be powers of 2

File: com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
+using System;
+
+namespace Unity.MLAgents.Sensors
+{
+    public class BufferSensor : ISensor
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs:5]: Inherits from IDimensionPropertiesSensor too, right?

File: com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        /// <param name="obs"> The float array observation</param>
+        public void AppendObservation(float[] obs)
+        {
+            if (m_CurrentNumObservables >= m_MaxNumObs)
+            {
+                return;
+            }
+            for (int i = 0; i < obs.Length; i++)
+            {
+                m_ObservationBuffer[m_CurrentNumObservables * m_MaxNumObs + i] = obs[i];
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs:40]:
```suggestion
                m_ObservationBuffer[m_CurrentNumObservables * m_ObsSize + i] = obs[i];
```
(I think)

File: com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            }
+            m_CurrentNumObservables++;
+        }
+
+        /// <inheritdoc/>
+        public int Write(ObservationWriter writer)
+        {
+            for (int i = 0; i < m_ObsSize * m_MaxNumObs; i++)
+            {
+                writer[i] = m_ObservationBuffer[i];
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs:50]: So this is implicitly zero padded, right?

File: com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs [hunk 5]
```diff
@@ -375,6 +375,18 @@ public static ObservationProto GetObservationProto(this ISensor sensor, Observat
                 {
                     observationProto.CompressedChannelMapping.AddRange(compressibleSensor.GetCompressedChannelMapping());
                 }
+                var dimensionPropertySensor = sensor as IDimensionPropertiesSensor;
+                if (dimensionPropertySensor != null)
+                {
+                    var dimensionProperties = dimensionPropertySensor.GetDimensionProperties();
+                    int[] intDimensionProperties = new int[dimensionProperties.Length];
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs:382]: nit: can you just loop and call `observationProto.DimensionProperties.Add()`?

File: com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs (hunk 4 above)
Comment (vincentpierre) [reply com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs:50]: Yes, the assumption is that if all the floats in an obs are 0, then the observation is padded. This padding will be ignored by the trainers, but still fed as input during inference / training. 
We can pad with another value, but I chose 0.

File: com.unity.ml-agents/Runtime/Sensors/IDimensionPropertiesSensor.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+{
+
+    /// <summary>
+    /// The Dimension property flags of the observations
+    /// </summary>
+    [System.Flags]
+    public enum DimensionProperty
+    {
+        /// <summary>
+        ///     No properties specified.
```
Comment (dongruoping) [inline com.unity.ml-agents/Runtime/Sensors/IDimensionPropertiesSensor.cs:11]:
```suggestion
        /// No properties specified.
```

File: com.unity.ml-agents/Runtime/Sensors/IDimensionPropertiesSensor.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        /// </summary>
+        Unspecified = 0,
+
+        /// <summary>
+        /// No Property of the observation in that dimension. Observation can be processed with
+        /// Fully connected networks.
```
Comment (dongruoping) [inline com.unity.ml-agents/Runtime/Sensors/IDimensionPropertiesSensor.cs:17]:
```suggestion
        /// fully connected networks.
```

File: com.unity.ml-agents/Runtime/Sensors/IDimensionPropertiesSensor.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        /// </summary>
+        None = 1,
+
+        /// <summary>
+        /// Means it is possible to do a convolution in this dimension.
```
Comment (dongruoping) [inline com.unity.ml-agents/Runtime/Sensors/IDimensionPropertiesSensor.cs:22]: I wonder if it will be better by saying something like "suitable/meaningful to apply a convolution".

(It's possible to do a conv anyways.)

File: ml-agents-envs/mlagents_envs/base_env.py [hunk 9]
```diff
@@ -435,17 +436,61 @@ def create_discrete(discrete_branches: Tuple[int]) -> "ActionSpec":
...
+    """
+    TRANSLATIONAL_EQUIVARIANCE = 2
+    """
+    Means that there can be a variable number of observations in this dimension.
+    The observations are unordered.
+    """
+    VARIABLE_SIZE = 4
+
+
+class ObservationSpec(NamedTuple):
```
Comment (ervteng) [inline ml-agents-envs/mlagents_envs/base_env.py:461]: I'm wondering if rather than having the ObservationSpec hold multiple `List[item]`, we have the `BehaviorSpec` hold a `List[ObservationSpec]` and have each ObservationSpec hold the information for one obs. That way, when we go through and e.g. create the input processors based on the DimensionProperty and type, we can do something like:

```
for obs_spec in observation_specs:
    _shape = obs_spec.shape
    _properties = obs_spec.dimension_properties
     _type = obs_spec.type # TBD
```

Rather than 
```
for _shape, _properties, _type in zip(observation_spec.shapes, observation_spec.dimension_properties, observation_spec.types):
```

File: com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs [hunk 10]
```diff
@@ -0,0 +1,95 @@
+using System;
+
+namespace Unity.MLAgents.Sensors
+{
+    public class BufferSensor : ISensor, IDimensionPropertiesSensor
```
Comment (ervteng) [inline com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs:5]: (probably my own ignorance but) why do the sensors implement both interfaces when IDimensionPropertiesSensor inherits from ISensor?

Comment (chriselion) [reply com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs:5]: The original proposal was that IDimensionPropertiesSensor would be its own interface and not inherit from ISensor.

File: ml-agents-envs/mlagents_envs/base_env.py (hunk 9 above)
Comment (chriselion) [reply ml-agents-envs/mlagents_envs/base_env.py:461]: > we have the BehaviorSpec hold a List[ObservationSpec] and have each ObservationSpec hold the information for one obs. 

I think I like this better too.

File: com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs (hunk 2 above)
Comment (vincentpierre) [reply com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs]: Yes

File: com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs (hunk 3 above)
Comment (vincentpierre) [reply com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs]: You are right

File: com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs (hunk 10 above)
Comment (ervteng) [reply com.unity.ml-agents/Runtime/Sensors/BufferSensor.cs:5]: Ah I see, it has changed 👍

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 11]
```diff
@@ -24,7 +24,7 @@
 class NetworkBody(nn.Module):
     def __init__(
         self,
-        observation_shapes: List[Tuple[int, ...]],
+        observation_spec: List[ObservationSpec],
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/torch/networks.py]: Very nit: `observation_specs`

File: ml-agents-envs/mlagents_envs/base_env.py [hunk 12]
```diff
@@ -435,17 +436,53 @@ def create_discrete(discrete_branches: Tuple[int]) -> "ActionSpec":
...
-    - action_spec is an ActionSpec NamedTuple
+    - sensor_spec is a List of SensorSpec NamedTuple containing
+    information about the information of the Agent's observations such as their shapes.
+    The order of the SensorSpec is the same as the order of the observations of an
+    agent.
+    - action_spec is an ActionSpec NamedTuple.
     """
 
-    observation_shapes: List[Tuple]
+    sensor_spec: List[SensorSpec]
```
Comment (ervteng) [inline ml-agents-envs/mlagents_envs/base_env.py:485]:
```suggestion
    sensor_specs: List[SensorSpec]
```

Comment (vincentpierre) [reply ml-agents-envs/mlagents_envs/base_env.py:485]: okay, will change throughout.

File: ml-agents-envs/mlagents_envs/environment.py [hunk 13]
```diff
@@ -129,7 +129,7 @@ def _warn_csharp_base_capabilities(
         if not caps.baseRLCapabilities:
             logger.warning(
                 "WARNING: The Unity process is not running with the expected base Reinforcement Learning"
-                " capabilities. Please be sure upgrade the Unity Package to a version that is compatible with this "
+                " capabilities. Please be sure upgrade the Unity Package sensor_specsn that is compatible with this "
```
Comment (vincentpierre) [inline ml-agents-envs/mlagents_envs/environment.py:132]: I think I have a problem with my IDE, this should not have changed

File: ml-agents-envs/mlagents_envs/base_env.py (hunk 12 above)
Comment (vincentpierre) [reply ml-agents-envs/mlagents_envs/base_env.py]: Any other issues you see with naming or design ?

File: ml-agents-envs/mlagents_envs/tests/test_registry.py [hunk 14]
```diff
@@ -36,7 +36,7 @@ def test_basic_in_registry():
     for worker_id in range(2):
         assert BASIC_ID in registry
         env = registry[BASIC_ID].make(
-            base_port=6005, worker_id=worker_id, no_graphics=True
+            base_port=6002, worker_id=worker_id, no_graphics=True
```
Comment (ervteng) [inline ml-agents-envs/mlagents_envs/tests/test_registry.py:39]: just checking if this change was intentional

**Review by ervteng — APPROVED** [review]
Super useful PR, LGTM
